### PR TITLE
feat: Add 'IntoSeries'

### LINKED ISSUE #387: Add IntoSeries type hint

### PR DESCRIPTION
**Pull request by luke396**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #387 

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below.

I think I should keep learning from #365 to improve the current PR.

### CODE REVIEW

File: narwhals/translate.py [hunk 1]
```diff
@@ -254,7 +255,7 @@ def from_native(
 
 @overload
 def from_native(
-    native_object: Any,
+    native_object: IntoSeriesT,
```
Comment (luke396) [inline narwhals/translate.py:258]: Due to the following code in `altair`, should we keep the `Any` type and introduce a new one specifically for `IntoSeriesT`? @MarcoGorelli

```python
def _from_array_like(obj: Iterable[Any], /) -> list[Any]:
    try:
        ser = nw.from_native(obj, strict=True, series_only=True)
        return ser.to_list()
    except TypeError:
        return list(obj)
```
from https://github.com/vega/altair/blob/11078c2676a921cdfc44477ffe0d2a43609e0306/altair/utils/schemapi.py#L502-L507

**Review by MarcoGorelli — COMMENTED** [review]
thanks @luke396 !

sorry this took me a while to get to, I had a company retreat + holiday recently so had relatively little time

looks good, just got a little question

**Comment by luke396** [comment]
> looks good, just got a little question

Should we miss some questions here?

File: narwhals/typing.py [hunk 2]
```diff
@@ -47,11 +51,17 @@ def __dataframe__(self, *args: Any, **kwargs: Any) -> Any: ...
 Frame: TypeAlias = Union["DataFrame[Any]", "LazyFrame[Any]"]
 """Narwhals DataFrame or Narwhals LazyFrame"""
 
+IntoSeries: TypeAlias = Union[
+    "Series", "DataFrame[Any]", "LazyFrame[Any]", "NativeSeries"
+]
```
Comment (MarcoGorelli) [inline narwhals/typing.py:56]: sorry, the question didn't go for some reason

I wanted to ask, why include `"DataFrame[Any]", "LazyFrame[Any]"` here?

Comment (luke396) [reply narwhals/typing.py]: I don't recall any particular reason; I just defined it to pass all checks and seem to have forgotten to restrict it again.

Let's try how about remove them.

File: narwhals/stable/v1/__init__.py [hunk 3]
```diff
@@ -738,7 +739,7 @@ def from_native(
 
 @overload
 def from_native(
-    native_dataframe: Any,
+    native_dataframe: IntoSeriesT | Any,  # remain `Any` for downstream compatibility
```
Comment (luke396) [inline narwhals/stable/v1/__init__.py:742]: for https://github.com/vega/altair/blob/796649efd751a7b928ac4f81e192d0ae93f4a3d0/altair/utils/schemapi.py#L502C1-L507C25

```python
def _from_array_like(obj: Iterable[Any], /) -> list[Any]:
    try:
        ser = nw.from_native(obj, strict=True, series_only=True)
        return ser.to_list()
    except TypeError:
        return list(obj)
```

File: narwhals/typing.py (hunk 2 above)
Comment (luke396) [reply narwhals/typing.py]: Seems fine!

File: narwhals/typing.py [hunk 4]
```diff
@@ -29,6 +29,10 @@ def columns(self) -> Any: ...
 
         def join(self, *args: Any, **kwargs: Any) -> Any: ...
 
+    class NativeSeries(Protocol):
+        @property
+        def name(self) -> Any: ...
```
Comment (MarcoGorelli) [inline narwhals/typing.py:34]: not sure this is true in general unfortunately 😭 
```python
In [4]: pa.chunked_array([['a', 'b', 'c']]).name
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 pa.chunked_array([['a', 'b', 'c']]).name

AttributeError: 'pyarrow.lib.ChunkedArray' object has no attribute 'name'

```

let's just put `__len__`?

**Review by MarcoGorelli — APPROVED** [review]
thanks @luke396 !

File: narwhals/typing.py (hunk 4 above)
Comment (luke396) [reply narwhals/typing.py]: Thank you for the correction. I didn't consider the situation.

**Pull request merged**